### DISCUSSION TITLE
o Question about how to distinguish objects in highlighting

### DISCUSSION THREAD

**Discussion by katawful** [tree-sitter/tree-sitter#1183] [Q&A] [answered]
I am working on syntax highlighting for my parser, and part of my issue is the fact that this language has a lot of built-in functions (~3000) as its a game scripting language. I would like to be able to make some ambiguous situations more clear in syntax highlighting, but I'm confused if its something for tree-sitter to handle or if its an editor level solution

My ambiguities:

```
namespace.variable
object.builtin_func

set x to builtin_func
set x to variable
```

**Comment by ahelwer** [comment]
Take a look at tree-sitter's [conflicting tokens rules](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens), specifically rule 4 about match specificity. If you have something like:
```
set_id: $ => seq(
  'set', $.id, 'to', choice($.variable, $.builtin_func)
),

$.id: $ => /\w+/,
$.variable: $ => /\w+/,
$.builtin_func: $ => choice('max', 'min', 'add, 'remove')
```
then tree-sitter will prefer matching the `set_id` rule to `builtin_func` because the match is more specific than the catch-all `variable` rule; you should then be able to use this to distinguish them for highlighting purposes.

**Reply by katawful** [reply]
My issue is that there's a lot of these built in functions (since it's for a game), which leads the grammar file to be too large to work properly. Last time I tried it took 30 minutes and about 15GB of memory to compile and then didn't even work properly

Is there another method within tree-sitter that lets me distinguish these or is it the job of the editor that implements it (like NeoVim's tree-sitter API)?

**Reply by ahelwer** [reply]
Huh, that's really strange behavior. Might be related to a couple of the other bugs around extreme memory usage, like https://github.com/tree-sitter/tree-sitter/issues/656. Perhaps you could work around it with an external scanner? You could even do something fun like make a prefix tree for fast matching! Although with 3000 elements I don't know whether the overhead would be worth it over a simple hardcoded table.

**Reply by katawful** [reply]
I could potentially work around it with a scanner, but I'm not the best at C/C++ and I'm not sure if it'll remain speedy still

**Reply by katawful** [reply]
I'm looking through some issues on the main page, and there's some similar problems as to mine. I will do some restructuring and testing to see if it's my grammar and not the size

**Comment by katawful** [comment]
So I restructured my grammar file in an attempt to simplify the generation, removing all conflicts in the process and while initially the generation took up very little memory with all the built-in functions, it still ended up thrashing my memory and failed to complete properly still. If I cut the number of functions down to ~250, it'll still consume large amounts of memory. Either this is a bug, or just not something tree-sitter can effectively do. Does anyone have any suggestions on how to achieve this problem?

**Reply by ahelwer** [reply]
Do you have a link to your grammar?

**Reply by katawful** [reply]
https://github.com/katawful/tree-sitter-obse/blob/builtin-function/grammar.js

Current version

**Reply by ahelwer** [reply]
Hmm, wonder whether the call to `caseInsensitive` is causing the blowup. Edit: nope still seems to take a while to generate (waiting for at least a few minutes now).

**Reply by katawful** [reply]
It still thrashes memory just the same when doing a smaller chunk

**Comment by ObserverOfTime** [comment, marked as answer]
You shouldn't hardcode 3000 built-in functions to the parser. That's too much even for queries.
